bson/bsoncodec/time_codec.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncodec
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson/bsonoptions"
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+const (
+	timeFormatString = "2006-01-02T15:04:05.999Z07:00"
+)
+
+var defaultTimeCodec = &TimeCodec{}
+
+// TimeCodec is the Codec used for struct values.
+type TimeCodec struct {
+	UseLocalTimeZone bool
+	DecodeFromString bool
+}
+
+var _ ValueCodec = &TimeCodec{}
+
+// NewTimeCodec returns a TimeCodec with options opts.
+func NewTimeCodec(opts ...*bsonoptions.TimeCodecOptions) (*TimeCodec, error) {
+	timeOpt := bsonoptions.MergeTimeCodecOptions(opts...)
+
+	codec := TimeCodec{}
+	if timeOpt.UseLocalTimeZone != nil {
+		codec.UseLocalTimeZone = *timeOpt.UseLocalTimeZone
+	}
+	if timeOpt.DecodeFromString != nil {
+		codec.DecodeFromString = *timeOpt.DecodeFromString
+	}
+	return &codec, nil
+}
+
+// DecodeValue is the ValueDecoderFunc for time.Time.
+func (tc *TimeCodec) DecodeValue(dc DecodeContext, vr bsonrw.ValueReader, val reflect.Value) error {
+	if !val.CanSet() || val.Type() != tTime {
+		return ValueDecoderError{Name: "TimeDecodeValue", Types: []reflect.Type{tTime}, Received: val}
+	}
+
+	var timeVal time.Time
+	valueType := vr.Type()
+	switch valueType {
+	case bsontype.DateTime:
```

Comment:
What about a bsontype.Int64? I believe we could easily consider that milliseconds from the epoch rather than erroring out.

Also, the bsontype.Timestamp should absolutely be allowed.

Reply:
I like the idea of adding it. Otherwise, it seems like we're just waiting for a new ticket about unmarshalling `bsontype.Timestamp` into a `time.Time`. I think we should just do it now.